ci/cron/src/Main.hs

```diff
@@ -0,0 +1,254 @@
+                        "---",
+                        err,
+                        "---"]
+die _ _ _ _ = Exit.die "Type system too weak."
+
+
+shell :: String -> IO String
+shell cmd = do
+    (exit, out, err) <- shell_exit_code cmd
+    if exit == Exit.ExitSuccess
+    then return out
+    else do
+        die cmd exit out err
+
+shell_ :: String -> IO ()
+shell_ cmd = do
+    _ <- shell cmd
+    return ()
+
+robustly_download_nix_packages :: IO ()
+robustly_download_nix_packages = do
+    h (10 :: Integer)
+    where
+        cmd = "nix-build nix -A tools -A cached"
+        h n = do
+            (exit, out, err) <- shell_exit_code cmd
+            case (exit, n, "unexpected end-of-file" `List.isInfixOf` err) of
+              (Exit.ExitSuccess, _, _) -> return ()
+              (_, 0, _) -> die cmd exit out err
+              (_, _, True) -> h (n - 1)
+              (_, _, False) -> die cmd exit out err
+
+http_get :: JSON.FromJSON a => String -> IO a
+http_get url = do
+    manager <- HTTP.newManager TLS.tlsManagerSettings
+    request' <- HTTP.parseRequest url
+    -- Be polite
+    let request = request' { HTTP.requestHeaders = [(Insensitive.mk (BS.fromString "User-Agent"), BS.fromString "DAML cron (team-daml-language@digitalasset.com)")] }
+    response <- HTTP.httpLbs request manager
+    let Network.HTTP.Types.Status.Status { Network.HTTP.Types.Status.statusCode = status } = HTTP.responseStatus response
+    let body = JSON.decode $ HTTP.responseBody response
+    case (status, body) of
+      (200, Just body) -> return body
+      _ -> Exit.die $ unlines ["GET \"" <> url <> "\" returned status code " <> show status <> ".",
+                               show $ HTTP.responseBody response]
+
+http_post :: String -> [(String, String)] -> LBS.ByteString -> IO LBS.ByteString
+http_post url headers body = do
+    manager <- HTTP.newManager TLS.tlsManagerSettings
+    request' <- HTTP.parseRequest url
+    -- Be polite
+    let request = request' { HTTP.requestHeaders = map (\(n, v) -> (Insensitive.mk (BS.fromString n), BS.fromString v))
+                                                       $ ("User-Agent", "DAML cron (team-daml-language@digitalasset.com)") : headers,
+                             HTTP.method = "POST",
+                             HTTP.requestBody = HTTP.RequestBodyBS $ BS.fromString $ LBS.toString body }
+    response <- HTTP.httpLbs request manager
+    let Network.HTTP.Types.Status.Status { Network.HTTP.Types.Status.statusCode = status } = HTTP.responseStatus response
+    case status `quot` 100 of
+      2 -> return $ HTTP.responseBody response
+      _ -> Exit.die $ "POST " <> url <> " failed with " <> show status <> "."
+
+github_versions :: [GitHubVersion] -> Set.Set String
+github_versions vs = Set.fromList $ map name vs
+
+remove_prereleases :: [GitHubVersion] -> [GitHubVersion]
+remove_prereleases = filter (\v -> prerelease v == False)
+
+docs_versions :: JSON.Value -> Set.Set String
+docs_versions json =
```

Comment:
Can we just accept a `[Text]` here and rely on the `FromJSON` instance to implement `to_list` at the callsite?

Reply:
I don't understand what you mean here.